### PR TITLE
Export/ulfm to ompi5

### PR DESCRIPTION
**Pull request by abouteiller**
Features
 ========
 This implementation conforms to the User Level Failure Mitigation (ULFM)
 MPI Standard draft proposal. The ULFM proposal is developed by the MPI
 Forum's Fault Tolerance Working Group to support the continued operation of
 MPI programs after crash (node failures) have impacted the execution. The key
 principle is that no MPI call (point-to-point, collective, RMA, IO, ...) can
 block indefinitely after a failure, but must either succeed or raise an MPI
 error.

 This implementation produces the three supplementary error codes and five
 supplementary interfaces defined in the communicator section of the
 [http://fault-tolerance.org/wp-content/uploads/2012/10/20170221-ft.pdf]
 (ULFM chapter) standard draft document.

 + `MPIX_ERR_PROC_FAILED` when a process failure prevents the completion of
   an MPI operation.
 + `MPIX_ERR_PROC_FAILED_PENDING` when a potential sender matching a non-blocking
   wildcard source receive has failed.
 + `MPIX_ERR_REVOKED` when one of the ranks in the application has invoked the
   `MPI_Comm_revoke` operation on the communicator.
 + `MPIX_Comm_revoke(MPI_Comm comm)` Interrupts any communication pending on
   the communicator at all ranks.
 + `MPIX_Comm_shrink(MPI_Comm comm, MPI_Comm* newcomm)` creates a new
   communicator where dead processes in comm were removed.
 + `MPIX_Comm_agree(MPI_Comm comm, int *flag)` performs a consensus (i.e. fault
   tolerant allreduce operation) on flag (with the operation bitwise AND).
 + `MPIX_Comm_failure_get_acked(MPI_Comm, MPI_Group*)` obtains the group of
   currently acknowledged failed processes.
 + `MPIX_Comm_failure_ack(MPI_Comm)` acknowledges that the application intends
   to ignore the effect of currently known failures on wildcard receive
   completions and agreement return values.

 ## Supported Systems
 There are several MPI engines available in Open MPI,
 notably, PML "ob1", "cm", "ucx", and MTL "ofi", "portals4", "psm2".
 At this point, only "ob1" is adapted to support fault tolerance.

 "ob1" uses BTL ("Byte Transfer Layer") components for each supported
 network. "ob1" supports a variety of networks that can be used in
 combination with each other. Collective operations (blocking and
 non-blocking) use an optimized implementation on top of "ob1".

 - Loopback (send-to-self)
 - TCP
 - UCT (InfiniBand)
 - uGNI (Cray Gemini, Aries)
 - Shared Memory (FT supported w/CMA and XPmem; KNEM is untested)
 - Tuned and non-blocking collective communications

More details available in README.FT.ULFM.md

### For p2p operations 

Each solid line represents an IMB-MPI1 run with Open MPI master. ucx/ib between ranks located at two different nodes, ucx/sm between two ranks located on sibling cores on a single node. The performance of pr7740 is overlayed with dashed lines and squares/circles. 

Left column is latency (lower is better); Right column is bandwidth (higher is better).

Graph rows alternate between the default setting on that machine (pml=ucx), for validation only, and  mode that supports runtime ft=on (pml=ob1/uct).

The patch introduces no substantive changes to the UCX pml pathway, so we expect no difference (which is what we see), and we cannot run with ft=on.

The PML ob1 pathway supports FT, and we see no performance difference with ft=off, and maybe a very slight effect with ft=on.

pr7740-p2p

### For coll operations

In this set of tests we run a selection of IMB collective benchmarks. We present only the pml=ob1 case (ucx case shows no difference). sm+cma is used for communication between cores, uct/mlx5 is used between nodes, map-by slot  (all other settings default, thus coll=tuned). 

 The results are a bit more variable as is illustrated by the fact that it happens that ft=on is faster in some instances than ft=off or master. We would need more statistics to cleanup the look of the curves, but overall the performance impact appears non-measurable. 

pr7740-coll

### Aries systems

This is a set of tests that compares ULFM with its upstream master  (an older version) on Cori (ugni). This graph aggregates a log of data as each violin point is the aggregate of hundred of IMB runs. Blue is master reference, orange is ulfm ft=on. We present the distribution of multiple runs per message size.

Again, the difference is minuscule, if any. 

UvO-pingpongbw

### CODE REVIEW

**Comment by rhc54** [comment]
yikes - can you squash this down?

**Comment by jjhursey** [comment]
It's my understanding that this PR is the OMPI side companion of https://github.com/openpmix/prrte/pull/542 in PRRTE.

**Comment by abouteiller** [comment]
bot:retest

**Comment by gpaulsen** [comment]
PRRTE has requested that the Open MPI community inform them if we're taking this for v5.0 or if we'll hold it for post v5.0.

Everyone, please review this PR, and we can discuss and make a decision on 5/26 web-ex.

**Comment by nathanweeks** [comment]
Is this PR still slated to make it into 5.0?

**Comment by bosilca** [comment]
yes. the conlicts will be removed asap.

**Review by devreal — DISMISSED** [review]
A couple of minor changes/comments/questions/missing braces. I cannot vouch for the conceptual correctness as I haven't tried to fully understand the algorithms (there is only so much time available) and I only quickly gazed over the build system and Fortran changes as I'm not familiar with either. Maybe someone else wants to look at these parts...

**Comment by abouteiller** [comment]
Thanks Joseph that's very thorough. I'll mark the items resolved as I go through it.

File: ompi/communicator/ft/comm_ft_detector.c
Comment (abouteiller) [reply ompi/communicator/ft/comm_ft_detector.c:147]: It does:  ERR_OUT_OF_RESOURCES is -2

File: ompi/communicator/ft/comm_ft_detector.c
Comment (abouteiller) [reply ompi/communicator/ft/comm_ft_detector.c:435]: In the normal case (fault-free) this is executed only once, and in any case not performance critical (executed after faults, once per fault).

File: ompi/communicator/ft/comm_ft_revoke.c
Comment (abouteiller) [reply ompi/communicator/ft/comm_ft_revoke.c:86]: Yes, but that needs to be done in a place where we hold the matching lock (i.e., it is done in mca_pml_ob1_revoke_comm)

File: ompi/include/mpi.h.in
Comment (abouteiller) [reply ompi/include/mpi.h.in:584]: Yes that's how autoconf populates these things.

File: ompi/communicator/ft/comm_ft.c
Comment (abouteiller) [reply ompi/communicator/ft/comm_ft.c:130]: You are right, that will also increase the SPC counter. However there's no internal ompi_wtime. Maybe it should be added, but not here.

File: ompi/communicator/ft/comm_ft_detector.c
Comment (devreal) [reply ompi/communicator/ft/comm_ft_detector.c:147]: Ahh yes! Yoda code misread, I have ^^

File: ompi/proc/proc.h
Comment (abouteiller) [reply ompi/proc/proc.h:80]: OpenSHMEM does use the padding, so it needs to remain available for that usage. We however definitely need to keep the field aligned for OpenSHMEM.

File: ompi/errhandler/errhandler.c
Comment (abouteiller) [reply ompi/errhandler/errhandler.c:368]: yes, yes there is, resolved.

File: ompi/mca/coll/base/coll_base_util.c
Comment (abouteiller) [reply ompi/mca/coll/base/coll_base_util.c:82]: Non FT errors may not be cancellable and could deadlock. What we need to do instead is verify if the error type is  cancellable, or free the request otherwise.

File: ompi/mca/pml/ob1/pml_ob1_isend.c
Comment (abouteiller) [reply ompi/mca/pml/ob1/pml_ob1_isend.c:166]: It is not safe. Working on a solution.

File: ompi/mca/coll/coll.h
Comment (abouteiller) [inline ompi/mca/coll/coll.h:624]: Possible module ABI compat issue @bwbarrett

File: ompi/mca/pml/pml.h
Comment (abouteiller) [inline ompi/mca/pml/pml.h:517]: module ABI change @bwbarrett

File: ompi/communicator/communicator.h
Comment (abouteiller) [inline ompi/communicator/communicator.h:222]: This is not an ABI change

File: ompi/proc/proc.h
Comment (abouteiller) [inline ompi/proc/proc.h:75]: Change in module ABI?

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/2d3651cde790e272420f592355401e0d

File: ompi/mca/coll/coll.h
Comment (abouteiller) [reply ompi/mca/coll/coll.h:624]: Sorry I had to force push to the branch because there was a problem with the way I did the squash initially. The links in these comments do not work anymore.

File: ompi/communicator/communicator.h
Comment (abouteiller) [inline ompi/communicator/communicator.h:222]: This change does not move existing fields, and due to padding in predefined should not cause ABI compatibility issues.

File: ompi/mca/pml/pml.h
Comment (abouteiller) [inline ompi/mca/pml/pml.h:518]: Adds a field to the module ABI in a conditional compilation. Make non-conditional?

File: ompi/mca/coll/coll.h
Comment (abouteiller) [inline ompi/mca/coll/coll.h:626]: Module ABI change in conditional compilation. Remove conditional?

File: ompi/proc/proc.h
Comment (abouteiller) [inline ompi/proc/proc.h:75]: Structure size/order dependent on compilation options. Remove conditional compilation?
This moves the proc_endpoints array. This could break ABI compat?

**Comment by jjhursey** [comment]
bot:ibm:pgi:retest

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/e2ef2b7a0a144795bec900fac8cfb535

**Comment by jjhursey** [comment]
The PGI failure output file is pretty long, but it's all there (you have to click through to see the full thing). It looks like a f90 issue with some of the new interfaces:
```
  PPFC     profile/pcomm_iagree_f08.lo
  PPFC     profile/pcomm_shrink_f08.lo
PGF90-F-0004-Unable to open MODULE file mpi_types.mod (comm_failure_get_acked_f08.F90: 15)
PGF90-F-0004-Unable to open MODULE file mpi_types.mod (comm_shrink_f08.F90: 15)
PGF90/power Linux 19.10-0: compilation aborted
PGF90/power Linux 19.10-0: compilation aborted
make[2]: *** [comm_failure_get_acked_f08.lo] Error 1
make[2]: *** Waiting for unfinished jobs....
PGF90-F-0004-Unable to open MODULE file mpi_types.mod (comm_agree_f08.F90: 15)
make[2]: *** [comm_shrink_f08.lo] Error 1
PGF90-F-0004-Unable to open MODULE file mpi_types.mod (comm_failure_ack_f08.F90: 15)
PGF90/power Linux 19.10-0: compilation aborted
PGF90-F-0004-Unable to open MODULE file mpi_types.mod (comm_iagree_f08.F90: 17)
PGF90-F-0004-Unable to open MODULE file mpi_types.mod (comm_revoke_f08.F90: 15)
PGF90/power Linux 19.10-0: compilation aborted
make[2]: *** [comm_agree_f08.lo] Error 1
PGF90/power Linux 19.10-0: compilation aborted
PGF90/power Linux 19.10-0: compilation aborted
make[2]: *** [comm_failure_ack_f08.lo] Error 1
make[2]: *** [comm_revoke_f08.lo] Error 1
make[2]: *** [comm_iagree_f08.lo] Error 1
```

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/0d6ffd2db3d1df246e634254cd85482e

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/476d7b6857d535d68a2c172c2c091be2

**Comment by nathanweeks** [comment]
Just checking on the status of this PR. Are additional testers needed?

**Comment by bosilca** [comment]
@nathanweeks any testing you can contribute on the resilience aspect would be more than welcomed.

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/9ee672eee7b1160d304dd4948f237c14

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/d636bd89cf5c13641c881fa46943c9c0

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/0bd7744a554a5e9f298db038a01de7b8

File: ompi/attribute/attribute_predefined.c
Comment (bwbarrett) [inline ompi/attribute/attribute_predefined.c:131]: I think having a conditional defined attribute is going to be a problem, due to the ordering bits in mpi.h.  Why not always create it, with a varying value?

File: ompi/communicator/communicator.h
Comment (bwbarrett) [reply ompi/communicator/communicator.h:222]: But given that we're already breaking ABI between 4.x and 5.0, should we add more padding?

File: ompi/include/mpi.h.in
Comment (bwbarrett) [inline ompi/include/mpi.h.in:197]: This makes me super nervous.  It seems like it would be much better to just always define the public bits of the FT library than to change the ABI based on whether or not FT is built?

File: ompi/include/mpif-values.pl
Comment (bwbarrett) [inline ompi/include/mpif-values.pl:245]: This is definitely problematic.  You can't add anything beyond MPI_FT in our current scheme, because the next thing may be 13 (FT enabled) or may be 12 (FT disabled).

File: ompi/include/mpif-values.pl
Comment (bwbarrett) [inline ompi/include/mpif-values.pl:335]: Same comment as above about the attribute.

File: ompi/mca/pml/pml.h
Comment (bwbarrett) [reply ompi/mca/pml/pml.h:518]: +1 :)

File: ompi/proc/proc.h
Comment (bwbarrett) [reply ompi/proc/proc.h:77]: I think I would just always leave this boolean in the proc structure and avoid any silly ABI questions.  That said, I would improve the comment so that it in some way references FT in the definition of "active".

File: ompi/communicator/communicator.h
Comment (abouteiller) [reply ompi/communicator/communicator.h:222]: There is 48 bytes left with LP64.

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/1c5d49907771e1fd8ad92fe3737af72d

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/b875bd150564e4a03b71b59ba437fc06

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/777e273b28cb62aed863a6a32be927a7

**Comment by abouteiller** [comment]
CI runtime failure is due to missing https://github.com/openpmix/prrte/pull/643 in prte.

**Comment by jsquyres** [comment]
Discussion on 5 Jan 2021 webex:

* It would be great to get this in to master as soon as possible.
    * We need time to stabilize this code in master.  So: the sooner, the better
    * Whether OMPI takes ULFM or not also has an impact on PMIx and/or PRRTE.  So we need to let our peer communities know what we're going to do.
* *Need* code reviews from people before this is merged.  You don't necessarily to have to review all the internals of ULFM (if you don't want to); emphasis should definitely be placed on making sure ULFM doesn't negatively non-ULFM OMPI/OSHMEM runs.
* This PR has been here for months.  If there are no meaningful issues by 26 Jan 2021, this PR can be merged.

**Comment by hppritcha** [comment]
bot:ompi:retest

**Comment by hppritcha** [comment]
I'm getting lots of warnings when I try to build this using gcc 4.8.5 on a RHEL7 system:
```
In file included from ../../../../../ompi/group/group.h:37:0,
                 from pgroup_compare_f.c:25:
../../../../../ompi/proc/proc.h:87:28: warning: redefinition of typedef 'ompi_proc_t' [-Wpedantic]
 typedef struct ompi_proc_t ompi_proc_t;
                            ^
In file included from ../../../../../ompi/mpi/fortran/mpif-h/prototypes_mpi.h:53:0,
                 from ../../../../../ompi/mpi/fortran/mpif-h/bindings.h:81,
                 from pgroup_compare_f.c:24:
../../../../../ompi/errhandler/errhandler.h:457:28: note: previous declaration of 'ompi_proc_t' was here
 typedef struct ompi_proc_t ompi_proc_t;
                            ^
In file included from ../../../../../ompi/group/group.h:37:0,
                 from pgroup_free_f.c:25:
../../../../../ompi/proc/proc.h:87:28: warning: redefinition of typedef 'ompi_proc_t' [-Wpedantic]
 typedef struct ompi_proc_t ompi_proc_t;
                            ^
In file included from ../../../../../ompi/mpi/fortran/mpif-h/prototypes_mpi.h:53:0,
                 from ../../../../../ompi/mpi/fortran/mpif-h/bindings.h:81,
                 from pgroup_free_f.c:24:
../../../../../ompi/errhandler/errhandler.h:457:28: note: previous declaration of 'ompi_proc_t' was here
 typedef struct ompi_proc_t ompi_proc_t;
                            ^
In file included from ../../../../../ompi/group/group.h:37:0,
                 from pgroup_excl_f.c:25:
../../../../../ompi/proc/proc.h:87:28: warning: redefinition of typedef 'ompi_proc_t' [-Wpedantic]
 typedef struct ompi_proc_t ompi_proc_t;
                            ^
In file included from ../../../../../ompi/mpi/fortran/mpif-h/prototypes_mpi.h:53:0,
                 from ../../../../../ompi/mpi/fortran/mpif-h/bindings.h:81,
```

**Comment by hppritcha** [comment]
Did some testing using ompi-tests/ibm and things look nominal.  Would be nice to get the compiler warnings fixed before merging this PR.  Also would be nice to decide whether to remove the --no-orte OMPI Jenkins CI test or else modify so that it passes before this PR is merged.

File: ompi/include/mpi.h.in
Comment (gpaulsen) [inline ompi/include/mpi.h.in:598]: I'd recommend removing lines 598 and 599 as unneeded.

**Comment by jjhursey** [comment]
Here is the CI error for the "--no-orte" (as mentioned this aliases to "--no-prrte"
```
configure:37519: ===== configuring 3rd-party/openpmix =====
configure:37708: running /bin/bash ./configure --disable-option-checking '--prefix=/home/ubuntu/workspace/open-mpi.build.autogen_options/AUTOGEN_OPTIONS/--no-orte/install' --without-tests-examples --disable-pmix-binaries --disable-pmix-backward-compatibility --disable-visibility --with-libevent=cobuild --with-prte-extra-lib=" /home/ubuntu/workspace/open-mpi.build.autogen_options/AUTOGEN_OPTIONS/--no-orte/src/3rd-party/libevent-2.1.12-stable/libevent.la /home/ubuntu/workspace/open-mpi.build.autogen_options/AUTOGEN_OPTIONS/--no-orte/src/3rd-party/libevent-2.1.12-stable/libevent_pthreads.la" 'CPPFLAGS= -I/home/ubuntu/workspace/open-mpi.build.autogen_options/AUTOGEN_OPTIONS/--no-orte/src/3rd-party/libevent-2.1.12-stable -I/home/ubuntu/workspace/open-mpi.build.autogen_options/AUTOGEN_OPTIONS/--no-orte/src/3rd-party/libevent-2.1.12-stable/include' --cache-file=/dev/null --srcdir=.
configure:37728: ===== done with 3rd-party/openpmix configure =====
configure:39687: error: Requested enabling fault-tolerance and using external launcher, but external PRTE doesn't support resilience; you can either use the internal PRTE, recompile the external PRTE with fault-tolerance, or disable fault-tolerance. Aborting.
```

**Comment by abouteiller** [comment]
> Did some testing using ompi-tests/ibm and things look nominal. Would be nice to get the compiler warnings fixed before merging this PR. Also would be nice to decide whether to remove the --no-orte OMPI Jenkins CI test or else modify so that it passes before this PR is merged.

Issue when using external prte has been resolved (ft will get compile-disabled if it was auto-enabled, error at configure is requested with explicit --with-ft configure flag).

The --no-orte flag is useful, but should be renamed --no-internal-prte


Launching with fault tolerance has been simplified (see prte https://github.com/openpmix/prrte/pull/726)

**Comment by rhc54** [comment]
> The --no-orte flag is useful, but should be renamed --no-internal-prte

There appears to be a misunderstanding of the `--no-orte` flag. I agree it should be renamed, but it probably should be changed to `--no-rte`. It specifically is meant to build OMPI without _any_ RTE to support environments that only want direct launch.

We specifically stated in telecons that OMPI did not want the ability to build against an external PRRTE, and @bwbarrett wrote the 3rd-party configure code accordingly. Since that time, the downstream packagers have indicated they plan to release PRRTE separately, so we might want to revisit that decision.

For now, however, `--no-orte` should completely disable the RTE, not link us against an external one.

**Review by bosilca — COMMENTED** [review]
I had few minor comments @abouteiller and myself need to address.

**Comment by abouteiller** [comment]
ready to go.

**Review by bosilca — APPROVED** [review]

**Pull request merged**
